Made day time flow variable

HenryLoenwind opened the pull request:
This lets mods control the ratio between game ticks and day ticks on a per-level level(*).

The changed ratio is also properly synced to clients, although vanilla clients will experience the sun rubber-banding/jumping once per second---but at least they still can connect (tested!).

This also adds a new command in 2 flavours:

```
/neoforge day speed set [default|realtime|<float>]
/neoforge day length set [default|realtime|<minutes>]
```

Where "default" disables the system, giving you the normal 20-minute day, and "realtime" sets the day length to 24 hours (just because it's trivial...). Leaving out the last parameter(s) will query the current value. This is available to all players, unlike the set commands that require the same access level as the `/gamerule` command (which can stop/start time).

For mods, the API entry point is `ServerLevel#setDayTimePerTick(float)`. Methods to query the value and to query the fractional part of the current day time are also available (also on client levels).

As Phantom spawning can be an issue with different day lengths than in vanilla, this PR also tweaks counting the time since a player last slept. When days are as long or shorter than in vanilla, Phantoms spawn based on _game_ ticks (as before). When days are longer, they spawn based on _day_ ticks (i.e. after 3 missed nights). Mods naturally can change that as they wish, they do have access to that stat anyway.

Help wanted: I'm not happy with the translations for the command feedback.

---

(*) For levels that do have independent daytime. Which is only the overworld in vanilla.

HenryLoenwind committed:
Made day time flow variable

marchermans commented on patches/net/minecraft/client/multiplayer/ClientLevel.java.patch:
```diff
@@ -150,5 +159,29 @@
++    @org.jetbrains.annotations.ApiStatus.Internal
++		public void setDayTimePerTick(float dayTimePerTick) {
++		    this.dayTimePerTick = dayTimePerTick;
++		}
```
This seems formatted weirdly.

marchermans commented on patches/net/minecraft/server/level/ServerLevel.java.patch:
```diff
@@ -257,5 +266,48 @@
++    	  serverLevelData.setDayTimePerTick(dayTimePerTick);
++        server.forceTimeSynchronization();
++    }
++
```
The formatting here is weird.

HenryLoenwind committed:
Merge remote-tracking branch 'neoforged/1.21.x' into time

HenryLoenwind commented:
About Phantoms:

These spawn based on game ticks, not day ticks. While this is fine for fast time flow---you wouldn't want phantoms every couple of seconds when fast-forwarding time---it is an issue for slow timeflow. If the day length is set to more than 2 hours, you can't even sleep enough to keep them at bay.

Should I do something about it, or should I leave it to mods to handle that using PlayerSpawnPhantomsEvent? And what about players who simply use the command and not a mod?

A possible solution would be to change the code in Player to check if the current level has custom time speed and if it's set to slower than normal, and only in this case to tick based on day ticks instead of game ticks. This would leave out the case when time flows too fast to sleep, but honestly, who leaves the game flickering like that for an hour anyway?

Opinions?

HenryLoenwind committed:
Fixed indentation in patches

and added safeguard against inadvertently stopping time with the API method (there's a game rule for that)

marchermans commented:
I would say to adapt the phantoms so that their spawn rate is adapted to the time of day.

Longer days, longer delay.

thedarkcolour commented:
Can day and night length be controlled separately? I used to really like this one 1.12 mod called Extended Days that did exactly this sort of thing, and you could make the days long and the nights short.

HenryLoenwind commented:
> Can day and night length be controlled separately? I used to really like this one 1.12 mod called Extended Days that did exactly this sort of thing, and you could make the days long and the nights short.

Not with the command, but it would be trivial for a mod to change the speed at dawn and dusk.

HenryLoenwind commented:
Setting to draft to handle the Phantoms...

Edit: Done

HenryLoenwind committed:
Slow down Phantom spawning if and only if day length is longer than vanilla

KnightMiner commented:
Is the command per dimension or will it set all dimensions globally? You mentioned it was level sensitive setter so I assume the command works per dimension.

Also, I assume by looking at the code that the value set by commands serialized so you don't need to constantly set it on world load. Is there a way to set it in a dimension added in a datapack though or will the datapack be forced to run a command to set it?

HenryLoenwind commented:
> Is the command per dimension or will it set all dimensions globally? You mentioned it was level sensitive setter so I assume the command works per dimension.
> 
> Also, I assume by looking at the code that the value set by commands serialized so you don't need to constantly set it on world load. Is there a way to set it in a dimension added in a datapack though or will the datapack be forced to run a command to set it?

Yes, it is per level (as long as that level has its own daytime---which in vanilla only the overworld has) and serialised. No, it can't be in a datapack (unless there's something that automatically runs commands?), you'd need at least a rudimentary mod...although that'd be a one-liner.

This is primarily intended for mods, I made the command to test it. There are a couple of issues you can run into when just using the command, like the "time since last sleep" running based on the dim you're in, not based on a dim you can sleep in. But IMHO that's stuff for a mod that can take over control and handle all kinds of stuff in any way it likes (i.e. there's more than one way to do it and neo shouldn't dictate how it's solved). I only added the basic Phantom control bit because it is very obvious and there's no way around that for someone using the command.
